feat: add restart from YAML option to Flow and Executors (#5488)
jina-ai/jina: docs/fundamentals/executor/executor-serve.md

```

(restart-executor)=
## Restart Executor in the Flow

Sometimes, you do not just change the Python files where the Executor logic is implemented, but you also want to change the YAML configuration of the Executor.
For this you can use the `restart` argument for the Executor in the Flow so that it watches changes in the configuration YAML code and makes sure that the Executor deployment is restarted with the new configuration.

In comparision with {ref}`reload <reload-executor>`, the Executor server is restarted with the new configuration YAML, in this case you can change the exact Executor class being used which is not possible with the {ref}`reload <reload-executor>` option.

````{admonition} Caution
:class: caution

> This feature is thought to let the developer iterate faster while developing but is not intended to be used in production environment.

Reviewer: ```suggestion
This feature aims to let developers iterate faster during development, but is not intended to for use in in production.
```

````

````{admonition} Note
:class: note
This feature requires watchfiles>=0.18 package to be installed.
````

> To see how this would work, let's have an Executor configuration defined in a file `executor.yml`.

Reviewer: ```suggestion
To see how this works, let's define an Executor configuration in the file `executor.yml`.
```

```yaml
jtype: MyExecutorBeforeRestart
```

We build a Flow with the Executor in it and expose it:

```python
import os
from jina import Flow, Executor, requests
os.environ['JINA_LOG_LEVEL'] = 'DEBUG'

class MyExecutorBeforeRestart(Executor):
   @requests
   def foo(self, docs, **kwargs):
       for doc in docs:
          doc.text = 'MyExecutorBeforeRestart'

class MyExecutorAfterRestart(Executor):
   @requests
   def foo(self, docs, **kwargs):
       for doc in docs:
          doc.text = 'MyExecutorAfterRestart'

f = Flow(port=12345).add(uses='executor.yml', restart=True)

with f:
    f.block()
```

We can see that the Executor is running and serving:

```python
from jina import Client, DocumentArray

c = Client(port=12345)

print(c.post(on='/', inputs=DocumentArray.empty(1))[0].text)
```

```text
MyExecutorBeforeRestart
```

We can edit the Executor YAML file and save the changes:

```yaml
jtype: MyExecutorAfterRestart
```
```

> You should see in the logs of the Flow

Reviewer: ```suggestion
In the Flow's logs we should see:
```
